ENH: Stats robust rebased

josef-pkt opened the pull request:
This is a rebased version of my old branch, updated to run with Python 3

https://github.com/statsmodels/statsmodels/compare/master...josef-pkt:stats_robust

see #6459

I moved the examples from `__main__` to the test module as a smoke tests.

pep8speaks commented:
Hello @josef-pkt! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`statsmodels/stats/robust_descriptive.py`](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/robust_descriptive.py):

> [Line 163:20](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/robust_descriptive.py#L163): [E999](https://duckduckgo.com/?q=pep8%20E999) SyntaxError: invalid syntax

* In the file [`statsmodels/stats/tests/test_robust_compare.py`](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/tests/test_robust_compare.py):

> [Line 13:1](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/tests/test_robust_compare.py#L13): [F401](https://duckduckgo.com/?q=pep8%20F401) 'statsmodels.stats.robust_compare.anova_oneway' imported but unused
> [Line 13:1](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/tests/test_robust_compare.py#L13): [F401](https://duckduckgo.com/?q=pep8%20F401) 'statsmodels.stats.robust_compare.anova_scale' imported but unused
> [Line 22:22](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/tests/test_robust_compare.py#L22): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'trim1'
> [Line 23:22](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/tests/test_robust_compare.py#L23): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'trim1'
> [Line 24:22](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/tests/test_robust_compare.py#L24): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'trim1'
> [Line 25:22](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/tests/test_robust_compare.py#L25): [F821](https://duckduckgo.com/?q=pep8%20F821) undefined name 'trim1'
> [Line 175:1](https://github.com/statsmodels/statsmodels/blob/5051fb8c49761118d5dc5b0b215c4ff0c2911cc3/statsmodels/stats/tests/test_robust_compare.py#L175): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

josef-pkt commented:
most likely there are still some bugs, e.g. df in anova_welsh looks wrong (too small in a simple example)

some quick checks with an unbalanced example
Welch t_test agrees with OLS with cov_type="HC2" except for the df that are not the same as in Satterthwaite, t-values are identical 

OLS nonrobust f_test and one-way anova completely agree

OLS HC2 f_test and anova_welch have large differences.

josef-pkt commented:
some problems in verifying the trim functions against R WRS2
I'm not sure whether I made mistakes or used different definitions.

var_winsorized is likely wrong (and might have mixed in computation for std_mean_trimmed

`tm.std_mean_trimmed` looks strange in WRS2

WRS2 uses
`np.sqrt(tm.var_winsorized / tm.nobs) / (1 - 2 * tm.fraction)`
while I also have the nobs correction implicitly  `(1 - 2 * tm.fraction)` inside the sqrt.
but I'm also using the actual number of non-trimmed observations.

another problem is desired fraction to cut is not the actual fraction to cut because of integer rounding/truncation

e.g. in the WRS2 vignette example, fraction to trim is 0.1, we trim one observation on each side which is 1/19

```
len(timevec), tm.nobs_reduced, tm.nobs * (1 - 2 * tm.fraction)
(19, 17, 15.2)
```

(AFAIR, I didn't look at WRS2 or the book of Wilcox when I wrote my versions, but I used several articles by Wilcox as reference)

The code also has a comment `# formula from an old SAS manual page, simplified`

I guess, I don't want to match WRS2 in some of the results.

The difference in `tm.std_mean_trimmed` between my version and WRS2 is large, so a MonteCarlo should show which is better.
(It's not a quick walk in the park)

josef-pkt commented:
:(
Nist dataplot also has the trim fraction outside of the sqrt
https://www.itl.nist.gov/div898/software/dataplot/refman2/auxillar/trimmse.htm
they reference Wilcox book

another one Medcalc
https://www.medcalc.org/manual/note-trimmedmean.php
This uses integer truncated fraction `k` and has corrected nobs `n - 2 k` in sqrts
reference to original 1963 article
(It looks like they also have the ddof the var_winsorized when it is used for std_trimmed_mean.
That looks the same as what I have in the repo version, but I changed it locally to have var_winsorized independent of the std_trimmed_mean use and matching WRS2)

josef-pkt commented:
R psych and WRS2 don't agree on winsorized statistics

```
 winsor.var(timevec, trim = 1/19)  # psych
[1] 137017.5818632456
> winvar(timevec, 1/19)  # WRS2
[1] 129679.0292397661
```

```
> winmean(timevec, 0.1)
[1] 380.1578947368421
> winmean(timevec, 1/19)
[1] 380.1578947368421

> winsor.mean(timevec, trim = 0.1)
[1] 326.0105263157894
> winsor.mean(timevec, trim = 1/19)
[1] 383.7340720221608
> winsor.mean(timevec, trim = 1/19 + 0.00001)
[1] 383.721841495845
> winsor.mean(timevec, trim = 1/19 - 0.00001)
[1] 383.7463025484766
```

Does psych do interpolation for fractional observations?

josef-pkt commented:
a quick monte carlo
my std_mean_trimmed underestimate the std a little bit 0.216 instead of 0.230  (around 6%)
larger sample size, nobs, doesn't make much difference for the underestimation (0.02199 / 0.0234 = -0.06 if nobs=1900)

And it's essentially the same if I use non-integer truncation fraction `TrimmedMean(x, 0.1)`, i.e. we use actual truncation on not desired fraction.

```
nobs = 19
n_rep = 10000
res = []
for i in range(n_rep):
    x = 1 + np.random.randn(nobs)
    tm_mc = TrimmedMean(x, 1/19)
    res.append([tm_mc.mean_trimmed, tm_mc.std_mean_trimmed])
    
res = np.asarray(res)
print(res.mean(0))
print(np.median(res, axis=0))
print(res.std(0, ddof=1))
[0.99653986 0.21589923]
[0.99973779 0.2148905 ]
[0.23064505 0.04012118]
```

If we want to do a "bias" correction, i.e. change df, then it would have to be proportional to nobs and nobs_reduced, i.e. correct something like 5 or 6 %

josef-pkt commented:
Ok, given that Wilcox version of sqrt works better in the MC:
equ. (3.9) p. 61 has the formula from the vignette appendix, and before that is the derivation that takes account of the correlation between trimmed observations. Because trimming is based on the order statistic, trimmed observations are not independent of each other anymore.

Wilcox, Rand R. 2012. Introduction to Robust Estimation and Hypothesis Testing. 3rd ed. Statistical Modeling and Decision Science. Amsterdam ; Boston: Academic Press.

josef-pkt commented:
aside: the `mean` function with trimming used in the WRS2 vignette, is from R base. ( I didn't find it in the WRS2 documentation)
https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/mean

another aside: Wilcox briefly mentions trimming based on interpolation. There was also an old R-help question why R base mean trimming doesn't interpolate.

I will stick with integer trimming, that's simpler, and used in most references that I looked at a long time ago (AFAIR)

josef-pkt commented:
I don't have a trimmed variance function, and neither does WRS2.
`var` is in R stats and does not have trimming

chemometrics in R has
sd_trim  https://rdrr.io/cran/chemometrics/man/sd_trim.html
They also have a correction factor to make it consistent with normal distribution. I guess similar to `mad` and `iqr`. 
I might have seen that before in robust estimation (but maybe that was based on statistical trimming and not on fixed percentage trimming. maybe in another ancient PR of mine, ELTS)

WRS2 lists under the `trimse` header the following functions

mean(self, 0.1) ## .10 trimmed mean
trimse(self, 0.1) ## se trimmed mean
winmean(self, 0.1) ## Winsorized mean (.10 Winsorizing amount)
winse(self, 0.1) ## se Winsorized mean
winvar(self, 0.1) ## Winsorized variance
median(self) ## median
msmedse(self) ## se median
mest(self) ## Huber M-estimator
mestse(self)
onestep(self) ## one-step M-estimator
mom(self) ## modified one-step M-estimator

Huber are in statsmodels.robust
I don't know about the median functions, we (including scipy.stats) don't have std_median/se_median
trimming and winsorizing are in this PR

josef-pkt commented:
and another problem, standard error of winsorized mean has currently a term under sqrt, that WRS2 has outside

```
std_mean_winsorized(self):
       ...
        # formula from an old SAS manual page, simplified
        std_ = np.sqrt(tm.var_winsorized / (tm.nobs_reduced - 1) *
                       (tm.nobs - 1.) / tm.nobs)
```
to match WRS2, I need
`np.sqrt(tm.var_winsorized / tm.nobs) * (tm.nobs - 1) / (tm.nobs_reduced - 1)`

What's "an old SAS manual page, simplified"?
:)

josef-pkt commented:
chasing  R functions and packages is a bit tiresome

One sample trimmed t-test
WRS2 doesn't have it, only for 2 and more sample comparisons, 
DescTools has it but results don't match up
finally PairedData package has also yuen test that works for one and two samples

```
> library(PairedData)
> yuen.t.test(timevec, y = NULL, tr = 1/19)

        One sample Yuen test, trim=0.0526315789473684

data:  x
t = 3.7115798169494, df = 16, p-value = 0.00189544440273
alternative hypothesis: true trimmed means is not equal to 0
95 percent confidence interval:
 146.9660486690165 538.4457160368659
sample estimates:
trimmed mean of x 
342.7058823529412 
```

josef-pkt commented:
and it continues with anova problems
it looks like WRS2 and PairedData which largely copied from Wilcox) only have Welsh-Yuen, but no Yuen with same variance assumption.

running my anova_welch, I get the same Satterthwaite df as WRS2, but the test statistic and p-value differ :(

Also, AFAICS, in the vignette appendix Yuen/Welch standard deviation for the mean diff don't just use the std of individual trimmed means, 
The correction factor from the one sample case is different or doesn't show up.

aside: the example that I use from the WRS2 docs is balanced, so the only difference between Welch and standard anova is in the degrees of freedom.

josef-pkt commented:
R is sooo nice, I found another package
https://www.rdocumentation.org/packages/onewaytests/versions/2.4/topics/welch.test

My welch test, agrees (at 1e-16) with those R packages when there is not trimming, also in unbalance sample case.

The 3 R packages agree with each other when there is trimming.

josef-pkt commented:
what's the difference between anova_bfm and anova_welch?
They produce the same result..
The scale/dispersion conversion is handled in anova_scale. Is anova_bfm just anova_welch?
(That would be logical, variance anovas like levene just use transformed data with regular anova.)

```
anova_welch([y0, y1[:-2]], trim_frac=1/12)
(2.2698412698412698, 0.15097418919782005, (1.0, 16.363527692920712))
anova_bfm([y0, y1[:-2]], trim_frac=1/12)
(2.2698412698412698,
 0.15097418919782002,
 0.15097418919782002,
 (1, 16.363527692920705, 1.0))

anova_welch([y0, y1[:-2]], trim_frac=0)
(2.1725749902610048, 0.15806721017979877, (1.0, 17.686748013960173))

anova_bfm([y0, y1[:-2]], trim_frac=0)
(2.1725749902610048,
 0.15806721017979877,
 0.15806721017979877,
 (1, 17.686748013960173, 1.0000000000000004))
```

However, levene with trimmed mean as center, doesn't do any further trimming and uses standard anova. In Yuen, trimmed anova we use the properties of the trimming and order statistics to compute the std_mean_trimmed (e.g. use var_winsorized and not regular var).

I guess, using levene with trimmed mean and trimming in anova does double trimming. 
Once to get the center for the data transformation `z =abs(x - mean_trimmed_x)` and then standard trimmed anova on `z`, i.e. we throw away the largest and smallest `z`.

`anova_scale` does not do the double trimming. It only optionally uses the mean_trimmed as center but then calls the untrimmed anova. 
(We could add a second trim_frac keyword for double trimming in anova_scale.)

Maybe:
anova_bfm has this sentence
"This currently returns both Brown-Forsythe and Mehrotra adjusted pvalues."
At least part of the code of bfm and welch differ, so it's not completly copy-paste.
So maybe it's just that BF used standard anova and mehrotra extended it to use Welch anova.

And we haven't even added Kruskal yet in the mix, i.e. rank_transformed data.

josef-pkt commented:
more info on Welsh ANOVA:
This is not just heteroscedasticity, i.e. compute mean under unequal variance assumption.
There is also a correction term in the statistic to give a better approximation to F distribution, 
df correction in Satterthwaite style is extra

OLS HC2 agrees with welch t-test and Welch Anova in two sample case
OLS HC2 test statistic agrees with welch anova without the extra correction term, when there are 3 or more samples in the comparison. 

Welsh 1950 refences James 1951 which has higher order expansion and correction for the heteroscedastic ANOVA. Welch corresponds to the first higher order expansion.

the following article compares bootstrap with Welch and James second order expansion
Krishnamoorthy, K., Fei Lu, and Thomas Mathew. 2007. “A Parametric Bootstrap Approach for ANOVA with Unequal Variances: Fixed and Random Models.” Computational Statistics & Data Analysis 51 (12): 5731–42. https://doi.org/10.1016/j.csda.2006.09.039.

josef-pkt commented:
And now the question is whether the variation with different transformation (trimmed, variane/Levene-Welch, equivalence Wellek-Welch) use the same correction factor that james/Welch derived for the standard mean comparison.
???

At least my trimmed Yuen-Welch doesn't match up with Yuen-Welsh in R, so some correction factor differs.

(I will add an extra option to Welch type anova to compute the HC2 solution, i.e. unequal var, but no Welch/Satterthwaite correction.)

josef-pkt commented:
confusing
AFAICS my old `std_mean_winsorized` wasn't std of winsorized mean, it was the std that is supposed to be used in welsh two-sample t_test and anova.
That differs for some reason (?) from `std_mean_winsorized`

old example code compared ttest only for the no trimming case

```
        tm = tm.reset_fraction(0)
        import statsmodels.stats.weightstats as smws
        smws._tstat_generic(tm.mean_trimmed, 0, tm.std_mean_trimmed,
                            tm.nobs_reduced - 1,
                            alternative='two-sided', diff=3)
```

josef-pkt commented:
yuen t_test using scipy ttest from_stats (we don't have the same kind of generic t-test in statsmodels)

results match yuen t_test and Yuen-Welch anova in R

```
from scipy import stats
mean1, std1, nobs1 = (tm0.mean_trimmed, 
                      np.sqrt(tm0.var_winsorized * (tm0.nobs - 1) / (tm0.nobs_reduced - 1)),
                      tm0.nobs_reduced)
mean2, std2, nobs2 = (tm1.mean_trimmed, 
                      np.sqrt(tm1.var_winsorized * (tm1.nobs - 1) / (tm1.nobs_reduced - 1)),
                      tm1.nobs_reduced)
​
stats.ttest_ind_from_stats(mean1, std1, nobs1, mean2, std2, nobs2,
                           equal_var=False)
Ttest_indResult(statistic=-1.5701656036912026, pvalue=0.13443998726837827)
```

josef-pkt commented:
let the fun continue: Brown Forsythe 1974

? which one?

Brown, Morton B., and Alan B. Forsythe. "Robust tests for the equality of variances." Journal of the American Statistical Association 69, no. 346 (1974): 364-367.

Brown, M.B. and Forsythe, A.B., 1974. The small sample behavior of some statistics which test the equality of several means. Technometrics, 16(1), pp.129-132.

The first one is variation of Levene anova for variance/dispersion
The second one is similar to Welch, but without variance in the weights for averaging.
(this one might be a bit better than welch when individual variances are noisy)

josef-pkt commented:
so anova_bfm is anova for means in the version of BF 1974, with correction by Mehrotra 1997
and it's not the variance BF

In the 3-sample case bfm and welch are not the same

```
anova_welch(arrs)
(8.023552121039247, 0.002114236255180824, (2.0, 24.272320628139))

anova_bfm(arrs)
(7.109006064211818,
 0.002838419657912239,
 0.0034295477871640816,
 (2, 31.42072561050516, 1.8555691977941184))

anova_oneway(arrs)
(7.474031933490758, 0.001746433041198707, (2, 40.0))
```

compared to R, anova_oneway and welch agree
```
> oneway.test(y ~ g, df3, var.equal = FALSE)

        One-way analysis of means (not assuming equal variances)

data:  y and g
F = 8.0235521210392, num df = 2.000000000000, denom df = 24.272320628139, p-value = 0.002114236255181

> oneway.test(y ~ g, df3, var.equal = TRUE)

        One-way analysis of means

data:  y and g
F = 7.4740319334908, num df = 2, denom df = 40, p-value = 0.001746433041199
```

bfm looks like exactly what the docstring says :)

I guess I was mixing up BF 1974 with BF 1974 when I started to look at this again.
but why does anova_scale call bfm by default, if that's the variance BF

```
    if method == 'bfm':
        res = anova_bfm(xxd)
```

josef-pkt commented:
quick reading of Mehrotra: The only difference is in one of the df which has a more complicated expression than in BF. Mehrotra cites the Technometrics BF 1974

code and docstring don't agree on which is called "denom" and which is called "num" (I always mix them up)

from above:
df BF (2,  31.42072561050516)
df M (1.8555691977941184, 31.42072561050516)

josef-pkt commented:
nice: R Onewaytests has BF in Mehrotra version, same results as mine

```
> bf.test(y ~ g, df3)

  Brown-Forsythe Test (alpha = 0.05) 
------------------------------------------------------------- 
  data : y and g 

  statistic  : 7.109006064211818 
  num df     : 2 
  denom df   : 31.42072561050515 
  p.value    : 0.002838419657912243 

  Result     : Difference is statistically significant. 
------------------------------------------------------------- 
```

and a good overview with formulas in the R journal article
Dag, Osman, Anil Dolgun, and Naime Meric Konar. 2018. “Onewaytests: An R Package for One-Way Tests in Independent Groups Designs.” The R Journal 10 (1): 175–99.
https://journal.r-project.org/archive/2018/RJ-2018-022/

the following article has an overview and formulas for various oneway anovas applied to levene-bf equality of variance based on abs diff from median
with MonteCarlo
Parra-Frutos, I. 2013. “Testing Homogeneity of Variances with Unequal Sample Sizes.” Computational Statistics 28 (3): 1269–97. https://doi.org/10.1007/s00180-012-0353-x.

josef-pkt commented:
oneway test for equality of variances, levene-BF

google search doesn't find a Levene-BF-Welch version
package lawstats has Levene-BF with zero-correction and bootstrap
package car has basic levene, I guess similar to scipy.stats.

termology naming  `anova_scale(... method="levene")` 
`method` chooses the anova method which should be "oneway" or something to indicate plain oneway under equal var assumptions.
levene would be standard oneway with center='mean'

I have an `exp(abs(x - m))` transformation for anova_scale. 
Which reference does it come from, and should it be `log(abs(x - m))`?

aside: one reference mentions levene-BF is test for "mad", which doesn't coincide with variance in non-normal distribution cases.
For normal distrribution. we could get the scaled mad equivalent to variance.

aside: other tests 
Fligner-Killeen Test Of Homogeneity Of Variances (uses ranks of abs(x-m)) and 
O-brien transform (available in scipy.stats)

josef-pkt commented:
default and recommendation, Barra-Frutos 2013

basic methods (without zero removal and without empirical/simulated p-values)
with Keyes-Levy, KL, small sample adjustment for effect of sample size on MAD.

Welch, James 1 and 2, and Alexander and Govern
- very liberal for very unequal sample sizes
- very conservative in very small samples (e.g. 5,5,5,5), but already good with 6,6,6,6
- a bit liberal with exponential distribution

Levene with median, BF, and BFM (Barra-Frutos uses BFR and cites Rubin before Mehrotra)
- very conservative in very small samples (e.g. 5,5,5,5), but already good with 6,6,6,6
- a bit conservative (0.025) at some smaller sample sizes
- never liberal ( > 0.075) in Barra-Frutos cases 
- BFR/BFM has a bit larger rejection rates  than Lev-med and BF, e.g. former slightly above 0.05, latter slightly below 0.05.

with empirical/simulated p-values and zero-removal (Noguchi-Gel, NG), all methods work pretty well

using NG, increases rejection rate. It removes underrejection in the conservative cases for L-med, BF, BFR, but makes it liberal in cases where the tests are close to correctly sized without zero-removal (but with KL)

They don't simulate cases with neither KL and NG, because expected values of the variance statistic depends on sample size, and will not be the same across groups with different nobs, even under the null of equal underlying variance.

josef-pkt commented:
API

- add `df` option that has effect in two cases, Welch vs HC2 and BF vs BFRM
- I'm not sure yet whether we want a shortcut `method` option that goes by names,
  eg. 
Welch = unequal var + welch_correction + df satterthwaite + (trim_frac=0)
HC0 = unequal var + (welch_correction=False) + df not satterthwaite
yuen = trimmed mean  + standard anova
yw = yuen-welsh = trimmed mean + all of welch

too messy for more combinations
anova_scale with (center=trimmed mean) + (transform=abs) + one of the oneways 

None of the functions in R packages contain more than a few cases. They either don't have other cases or have them in separate functions.

maybe just docstring by deviation from standard case

e.g. 
- equal variance assumptions
  - homogeneous standard f-oneway
  - heterogeneous
    - Welch
    - BF
    - not yet James, ...
- approximation
  - welch correction
  - satterthwaite df, changes  df_denom (second, lower df)
  - BF versus RB df, changes df_num (first, upper df)
  - use_f ? chisquare versus f distribution. Do we need that?
- corrections in anova_scale
  - zeros (not for now)
  - KL/NG

However, not all combinations make sense and shouldn't be allowed or warn or are ignored

equal variance assumptions could be put in `use_var` option, e.g. use_var can be in {"equal", "welch", "bf"}, with later extensions to "james" or others.
instead of binary `var_equal` in scipy.stats or some R functions.
(this sounds better to me) 

`anova_scale` can get a `kwds_oneway` or `**kwds` option, that are transmitted to the underlying anova

josef-pkt commented:
Too many options (like RLM), so the default will be used most of the time.
I guess, I prefer BFRM as default, even if it won't be the most familiar.

AFAICS, users can switch with just `use_var="welch"` or `use_var="equal"`

not decided yet: `Returns`
We need a results instance, to get more information from the oneway comparison. But I guess just a bit more than the basics. 
Not clear to me: can we get effect size that differs across methods? Or maybe just equal versus unequal variance?

In two sample case we could add conf_int and other things. But that ends up doing a TwoSampleComparison class, or the analogy of the t_test results class in models.

Can we reuse or combine with `wald_test` results class(es)?

josef-pkt commented:
checking with Stata

Stata does not have Welch, unequal variance option in `oneway`
`ttest` has 3 options for unpaired data, equal var, `unequal` and `welch`
AFAICS, it's not explicitly mentioned in docs, `unequal` doesn't use Satt df, `welch` does

josef-pkt commented:
another API idea:

add `groups=None`
if groups is None, then we assume data is list of array_likes
if groups is a 1-d array_like, then we take it as the group index and assume data is in long format.

`oneway([y1, y2, y3], groups=None, ....)`
`oneway(df["y"], groups=df["groups"], ...., use_var="welch", ...)`

If `oneway` were a class, then we could add `from_xxx`, `from_long`, `from_formula`

josef-pkt commented on statsmodels/stats/robust_compare.py:
```diff
@@ -0,0 +1,534 @@
+    nobs_t = nobs.sum()
+    n_groups = len(means)
+    # mean_t = (nobs * means).sum() / nobs_t
+    if use_var == "separate":
```
silly bug, wrong keyword value "unequal"

josef-pkt commented on statsmodels/stats/oneway.py:
```diff
@@ -0,0 +1,272 @@
+    use_var : {"unequal", "equal"}
+        If ``use_var`` is "unequal", then the variances can differe across
+        samples and the effect size for Welch anova will be computed.
+    ddof_between : int
```
I think `ddof_between` can be dropped.
(I might have interpreted things incorrectly when I rearranged the way things are computed)

josef-pkt commented on statsmodels/stats/oneway.py:
```diff
@@ -0,0 +1,377 @@
+    # Not sure if I need it in this function
+    compute_df = False
+    welch_correction = False
+    if compute_df:
```
no test coverage for this part
delete, and copy paste from anova_generic if we need it later

josef-pkt commented:
scale/variance anova code in robust_compare.py doesn't have much/any unit test coverage

josef-pkt commented:
what's the status? It has been some time

```
import statsmodels.stats.oneway as sma
import statsmodels.stats.robust_compare as smr
```

from dir(sma):

| sma | comment |
| --- | --- |
|'Holder', | |
| '_fstat2effectsize', | |
| 'anova_generic', | |
| 'confint_effectsize_oneway', | |
| 'confint_noncentrality', | |
| 'convert_effectsize_fsqu', | |
| 'effectsize_oneway', | |
| 'oneway_equivalence_generic', | |
| 'power_oneway_equivalence', | |


from dir(smr):

| smr | comment |
| --- | --- |
|  'TrimmedMean', | |
| 'anova_bfm' | obsolete |
| 'anova_oneway', | obsolete |
| 'anova_scale', | move |
| 'anova_welch', | obsolete |
|  'scale_transform', | move |
| 'trim_mean', | |
|  'trimboth' | |

This PR also includes `robust_descriptive.py` which I haven't looked at again.

josef-pkt commented:
Satterthwaite/Welch df correction looks strange, df looks much too small 
(still incompletely unit tested)

larger sample, nobs = [500, 600, 650, 750]

```
 use_var=unequal, welch_correction=True, trim_frac=0.00
(10.02089494882344, 1.5663673679465867e-06) (3.0, 1327.507631274335)

 use_var=unequal, welch_correction=False, trim_frac=0.00
(10.030959819977719, 1.5441239871989664e-06) (3.0, 1327.507631274335)

 use_var=equal, welch_correction=True, trim_frac=0.00
(10.288033108797187, 0.5118943480438943) (3.0, 0.33719931256276525)

 use_var=equal, welch_correction=False, trim_frac=0.00
(10.288033108797187, 0.5118943480438943) (3.0, 0.33719931256276525)

 use_var=bf, welch_correction=True, trim_frac=0.00
(10.210148443761362, 1.1493771688064426e-06) (2.9913948361408758, 2361.8283348362197)

 use_var=bf, welch_correction=False, trim_frac=0.00
(10.210148443761362, 1.1493771688064426e-06) (2.9913948361408758, 2361.8283348362197)
```

josef-pkt commented:
a quick mc, 10,000 replication,
null case for standard anove, all observations are i.i.d. N(0, 1)

welch_correction has no effect if use_var is not "unequal"
`nobs = np.array([10, 12, 13, 15]) * k_repl`

In large sample size is very good.
In small sample size is pretty good, but `use_var=unequal, welch_correction=False` overrejects a bit, i.e. using welch correction is better.

smaller sample
```
k_repl = 1

[0.008 0.022 0.045 0.107 0.191] use_var=unequal, welch_correction=True, trim_frac=0.00
[0.011 0.03  0.06  0.117 0.212] use_var=unequal, welch_correction=False, trim_frac=0.00
[0.007 0.03  0.052 0.104 0.187] use_var=equal, welch_correction=True, trim_frac=0.00
[0.007 0.03  0.052 0.104 0.187] use_var=equal, welch_correction=False, trim_frac=0.00
[0.006 0.024 0.047 0.097 0.183] use_var=bf, welch_correction=True, trim_frac=0.00
[0.006 0.024 0.047 0.097 0.183] use_var=bf, welch_correction=False, trim_frac=0.00

[0.008 0.022 0.048 0.108 0.209] use_var=unequal, welch_correction=True, trim_frac=0.10
[0.01  0.031 0.063 0.13  0.231] use_var=unequal, welch_correction=False, trim_frac=0.10
[0.007 0.018 0.046 0.1   0.193] use_var=equal, welch_correction=True, trim_frac=0.10
[0.007 0.018 0.046 0.1   0.193] use_var=equal, welch_correction=False, trim_frac=0.10
[0.006 0.017 0.044 0.099 0.19 ] use_var=bf, welch_correction=True, trim_frac=0.10
[0.006 0.017 0.044 0.099 0.19 ] use_var=bf, welch_correction=False, trim_frac=0.10

[0.014 0.033 0.052 0.115 0.212] use_var=unequal, welch_correction=True, trim_frac=0.20
[0.018 0.04  0.066 0.142 0.241] use_var=unequal, welch_correction=False, trim_frac=0.20
[0.011 0.025 0.057 0.101 0.208] use_var=equal, welch_correction=True, trim_frac=0.20
[0.011 0.025 0.057 0.101 0.208] use_var=equal, welch_correction=False, trim_frac=0.20
[0.008 0.016 0.049 0.09  0.195] use_var=bf, welch_correction=True, trim_frac=0.20
[0.008 0.016 0.049 0.09  0.195] use_var=bf, welch_correction=False, trim_frac=0.20
```

larger sample, nobs is [500, 600, 650, 750]
```
k_repl = 50

[0.0115 0.025  0.0507 0.1033 0.2034] use_var=unequal, welch_correction=True, trim_frac=0.00
[0.0117 0.0251 0.0508 0.1036 0.2038] use_var=unequal, welch_correction=False, trim_frac=0.00
[0.0113 0.0242 0.0503 0.103  0.2035] use_var=equal, welch_correction=True, trim_frac=0.00
[0.0113 0.0242 0.0503 0.103  0.2035] use_var=equal, welch_correction=False, trim_frac=0.00
[0.0115 0.0244 0.0498 0.1031 0.203 ] use_var=bf, welch_correction=True, trim_frac=0.00
[0.0115 0.0244 0.0498 0.1031 0.203 ] use_var=bf, welch_correction=False, trim_frac=0.00

[0.0118 0.0268 0.05   0.0986 0.2043] use_var=unequal, welch_correction=True, trim_frac=0.10
[0.0118 0.0269 0.0503 0.0987 0.2047] use_var=unequal, welch_correction=False, trim_frac=0.10
[0.0114 0.0262 0.0509 0.0976 0.2042] use_var=equal, welch_correction=True, trim_frac=0.10
[0.0114 0.0262 0.0509 0.0976 0.2042] use_var=equal, welch_correction=False, trim_frac=0.10
[0.0113 0.026  0.0505 0.0973 0.2035] use_var=bf, welch_correction=True, trim_frac=0.10
[0.0113 0.026  0.0505 0.0973 0.2035] use_var=bf, welch_correction=False, trim_frac=0.10

[0.0108 0.027  0.0513 0.1016 0.2023] use_var=unequal, welch_correction=True, trim_frac=0.20
[0.011  0.0271 0.0516 0.1021 0.2027] use_var=unequal, welch_correction=False, trim_frac=0.20
[0.0104 0.027  0.0517 0.1007 0.2005] use_var=equal, welch_correction=True, trim_frac=0.20
[0.0104 0.027  0.0517 0.1007 0.2005] use_var=equal, welch_correction=False, trim_frac=0.20
[0.0104 0.027  0.0515 0.1005 0.2009] use_var=bf, welch_correction=True, trim_frac=0.20
[0.0104 0.027  0.0515 0.1005 0.2009] use_var=bf, welch_correction=False, trim_frac=0.20
```

josef-pkt commented:
MC using t(5) distribution, all obs iid
Welsh and BF look good, 
BF rejection rate stays often just below alpha, others can be slightly above size,
trimming has not much affect

josef-pkt commented:
Based on the Monte Carlo, oneway anova implementation looks good. Other parts still need improvements.

Interface for equivalence testing ?

that's currently two calls, we need anova to get the effect size measure (anova test statistic)
(based on the unit tests)

```
eps = 0.5
res0 = anova_generic(means, stds**2, nobs, use_var="equal")
f = res0.statistic
res = oneway_equivalence_generic(f, n_groups, nobs.sum(), eps,
                                 res0.df, alpha=0.05)
```

and Welch (note needs welch_correction=False to match article)
```
res0 = anova_generic(means, stds**2, nobs, use_var="unequal",
                             welch_correction=False)
f = res0.statistic
res = oneway_equivalence_generic(f, n_groups, nobs.sum(), eps,
                                         res0.df, alpha=0.05)
```

Need one or two interface/wrapper functions ? one from stats and one from data
Current unit tests include equivalence for standard anova and for welch anova
`# reference numbers from Jan and Shieh 2019, p. 6`

related: what's the api for effect size and non-centrality parameter?

Do I have or get confint for anova effect size or noncentrality?

Can I do consistency check between hypothesis test/confint and equivalence, similar to what I did for proportion tost?
Might be possible using "f", but for data/stats I would need worst case for equivalence. The latter might not even work if the equivalence test doesn't reach the alpha bound (size < alpha).

google search doesn't find an R package for Wellek Anova, 
stats.stackexchange also doesn't have much, e.g. https://stats.stackexchange.com/questions/445216/equivalence-testing-with-multiple-means-in-r

josef-pkt commented:
checking internal consistency of equivalence testing with MC

- simulate data with identical means, i.e. all obs are iid, small differences in nobs across samples
- equivalence margin `eps` is 10 times the critical value for effect size at alpha=0.05, 
  I found critical value as average of MC simulations
- 3 anova version for use_var `options_var = ["unequal", "equal", "bf"]`


average wellek effect size by anova version
```
f_mc.mean(0)
array([0.02423054, 0.02413395, 0.02413222])
```
rejection rate at 0.05, based on pvalue, and results attribute `reject` (bool)
```
reject_mc.mean(0), frac_rej[:, 2]
(array([0.2866, 0.2879, 0.287 ]), array([0.2866, 0.2879, 0.287 ]))
```

other are [res1.crit_f, res1.crit_es, res1.power_zero], anova versions in rows
Last column is power if means are equal, this is very close to MC rejection rate
Critical values for standard anova (2nd row) do not depend on data, i.e. constant across MC replication
```
other_mc.mean(0).reshape(3, -1), np.ptp(other_mc, axis=0).reshape(3, -1)
(array([[0.4538525 , 0.01089246, 0.28521105],
        [0.45431441, 0.01090355, 0.28563897],
        [0.4531944 , 0.01087667, 0.28644404]]),
 array([[5.85365397e-05, 1.40487695e-06, 5.41996928e-05],
        [0.00000000e+00, 0.00000000e+00, 0.00000000e+00],
        [1.10452906e-02, 2.65086975e-04, 8.16779419e-03]]))
```

if I pick eps equal to the average crit value at alpha=0.05, then MC rejection rates and power_zero are close to 0.05. Increasing eps, i.e. widening the equivalence margin,  increases the rejection rate and power.

Differences in effect size and critical values across anova versions are small in this "zero" case.
In the asymmetric variance case, there might be larger differences. For internal consistency, we might need Welch, BF specific equivalence margins.

This still uses Wellek's effect size and needs to be changed. Plus a recipe for how to compute equivalence margin eps from an "example dataset". I guess that's just a rescaled anova statistic.

related: `power_oneway_equivalence` still needs to be fixed, so it can be used for any effect size under the alternative.

josef-pkt commented:
A reminder: effect size estimate is not zero if means are equal. See discussion on the different versions of `eta` effect sizes.

We don't care about this here, effect sizes are defined in how they are used in the hypothesis tests.
e.g. in the example above, `eps = 0.0191` is the wellek effect size at which the equivalence test has the correct 0.05 size in the equal means, iid case.
That's like a minimum detectable equivalence effect size (at given alpha).
(not sure anymore; I think we had some "empty" set cases in proportions for tost power. I haven't looked if we get something like that in this case.)

josef-pkt commented:
`effectsize_oneway` also still needs work
It largely duplicates code of anova_generic, but that includes now also "bf".
also docstring says f2, but we return sqrt(f2)

I guess I should go back to delegating to anova_generic in spite of some overhead, unused computation, at least until this parts are matured.

example usecase: **power as a function of "overall" sample size.**

- for H0 compute `f2_null` based on mean, vars_ specifying the null hypothesis, e.g. equal means
  we need nobs ratio  (standard anova with equal var does not depend on nobs ration, Welch and BF do)
- for H1 compute `f2_alt` for a base model given by mean vars_ and some standardized nobs, i.e. nobs ratio or distribution across samples
- compute power for replications of "base" model.

question: Can we have **fractional nobs**? e.g. nobs_i = fraction of sample i in total nobs.
That should not be a problem for effect size computation, but is problematic for degrees of freedom if we reuse anova_generic (negative df?)

**power for trimmed** (Yuen) anova: Can we just use regular power computation and correct for trimming fraction? Looks likely but I didn't check any literature for this.

question relevant for applied work: If we use standard anova for effect size and power, then how wrong are power and sample size computation if variances are unequal and we use Welch or BF as hypothesis test instead of standard "equal" anova?

**usecase post-hoc power** for arbitrary effect size alternative, compute power given a current sample for different alternatives
- narrow case: empirical/observed power, compute power at observed sample effect size 
- general case: allow arbitrary alternatives either defined in terms of base statistics, mean_alt, vars_alt and actual nobs, or in terms of a given effect size `f2_alt`. `vars_alt` could also be equal to observed `vars_`

Observed power could be added to anova_generic. Power under alternative needs separate function. `oneway_equivalence_generic` includes power against an "obvious" alternative, i.e. equal means.

josef-pkt commented:
diff between standard anova and Welch in power

setup: unequal means, unequal variance
means = np.array([-1, 0, 0, 1]) * 0.5
vars_ = np.arange(1, len(means) + 1)

Welch has higher power than standard and BF anovas at 0.05

```
[0.01  0.025 0.05  0.1   0.2  ]
[0.7729 0.8656 0.9146 0.956  0.9811] use_var=unequal, trim_frac=0.00
[0.5957 0.7307 0.8258 0.9024 0.9535] use_var=equal  , trim_frac=0.00
[0.6064 0.7449 0.8378 0.9122 0.9625] use_var=bf     , trim_frac=0.00
```

effect size using `f_alt = sma.effectsize_oneway(means_alt, vars_, nobs, use_var=use_var)`
f_alt: equal: 0.2144939010503242,  "unequal": 0.24625218457569975]
welch has larger effects size under alternative

FTestAnovaPower using effect sizes
"equal"-effectsize: 0.8162055275986806
"unequal" welch-effectsize  0.9155083973083813

those are pretty close to MC power
nobs is [50, 60, 65, 75] so not a very small sample (for mean testing)

Aside: FTestAnovaPower has equal means as a hard coded Null, i.e. effectsize under null = 0
We cannot just use a diff= mean_alt - mean_null as in t- or z-test power to define effect size.
effect size for power is noncentrality / nobs.
FTestAnovaPower uses central F, i.e. noncentrality = 0 under the Null.

The power for the equivalence test uses noncentral F under both null and alternative, AFAIU.
Similar to one sided tests with a value.

I guess we will get more of this late, e.g. for noncentral t in standardized mean difference as effect size, and maybe if we get some chisquare and f-test with a nonzero `value`.
This is currently not available in our power classes.

josef-pkt commented:
I don't manage to get effect size and power for use_var="bf"
effect size computed based on the f-statistic for BF is too small. Power computed with FTestAnovaPower is too small. I guess the problem is in the different degrees of freedom, the computation of the statistic looks fine, but effect size won't be comparable to standard and Welch anova. Note, welch effect size doesn't include welch correction.

Even when all variances are the same, BF power computation is much too low, while standard and welch anova are the same.

Back to FTestAnovaPower: this hardcodes degrees of freedom for standard anova.
I guess there are two possibilities
- find an effect size definition that is equivalent to standard anova, so df are correct.
  In the Monte Carlo Welch power calculation using FTestAnovaPower is pretty close to simulations already.
- add options for an FTestXXXPower class to specify df as a function of nobs_total.

josef-pkt commented:
reverse engineering

BF effect size is missing a factor `(1. - nobs / nobs_t).sum()`.
The computation of the denominator in BF statistic uses a weighted average of vars that doesn't add up to 1, and this is the missing piece.

now,
- in balanced, equal variance case, all 3 effect sizes are the same.
- in case with equal variance but unbalanced sample sizes, all 3 effect sizes are also the same
- in unequal variance, unbalanced sample size case, effect size and power differ, equal the lowest, welch the highes, BF in between.

theoretical power, which is close to power in MC

```
unequal 0.6240871638365101
equal   0.4960471233010961
bf      0.5335997922502402
```

josef-pkt commented:
The power computation overestimate power for standard "equal" and for BF anova (by around 3% in example). 
Power for welch is very close.

"equal" is misspecified for data, 
"BF": FTestAnovaPower approximation is not very close to "real" power function, or I'm missing something for a better approximation.


unequal var, `vars_ = [1, 2, 3, 4]`
large sample, `nobs = [500, 600, 650, 750]`

FTestAnovaPower
```
unequal 0.2769267212300879
equal   0.21646898268951853
bf      0.2329607663298614
```

MC pretty large,, large number of replication k_mc = 50000

rejection frequency at alpha = 0.05
```
frac_rej[:, 2]
array([0.2746 , 0.1884 , 0.20454])
```

josef-pkt commented:
more reverse engineering for "BFM"
I'm adding a correction term for df_denum correction by M: `f2 *= df_num / df_num2`

Now I have power close to MC rejection frequency
FTestAnovaPower    0.2077756596056003
MC frac  0.20454 from above 50,000 replication, 0.2084 latest run with 10,000 replication
(I'm not yet setting seed for MC)

josef-pkt commented:
Since the power Monte Carlo worked well in this case, we might add a public function.
I opened #6716 for this.

I guess not in this PR because I don't want to figure out the API for another function.

The remaining work for `oneway` is to finish power for equivalence anova, API and options.

josef-pkt commented:
informative names again, I have a difficult time keeping track of what I was doing in my notebook.
e.g. 
`f` can be f-statistic from test or a variant of cohen's effect size
`eps` is margin of equivalence in terms of Wellek's effect size, and we have other eps at alternatives

`nobs` is also confusing because I use it for array of nobs of samples, there is no plural "numbers of observations",
in power `nobs` corresponds to `nobs_t = nobs.sum()`

For Welch and BF, we have two `f`, the corrected effect size in `effectssize` oneway, and one in terms of f-statistic / nobs_t from the test (which is the noncentrality parameter in terms of F- or ncF distribution in the hypothesis tests.

josef-pkt commented:
another confusion, f or f_squared (`f2`)

Most of the time we need f_squared. However, current FTestAnovaPower uses sqrt
`pow_ = stats.ncf.sf(crit, df_denom, df_num, effect_size**2 * nobs)`

`effectsize_oneway` returns f and not f_squared which took me a while to remember when results didn't match up.

changing it in `FTestAnovaPower` is not backwards compatible.

AFAIR/AFAIU We can use eta2 to get an interpretable margin. eta2 is fraction of explained variance
If we assume margin eta2=0.05, then we get an f2 margin a bit larger than 5%
```
sma.convert_effectsize_fsqu(f2=None, eta2=np.array([0.05, 0.1]))
f2 = [0.05263158 0.11111111]
eta2 = [0.05 0.1 ]
```

josef-pkt commented:
I'm standardizing on Cohen's fsquared `f2 = nc / nobs_t`  with or without small sample, Welch, BF corrections.
`equiv_margin` does not contain name of effect size, but should be fine if standard is f2
Equiv test retains `margin_type` for now just to be able to use standard Wellek anova from literature.

I'm planning to eventually replace FTestAnovaPower with FAnovaPower or similar named class that uses fsquared. (Maybe that one will have the proper df computation for Welch and BF)

josef-pkt commented:
I haven't started yet to clean up oneway-scale. It's still in robust_compare

josef-pkt commented:
A pain, two hours debugging because I had forgotten about a typo in a NCSS dataset.

I have some test cases in 1 or 2 month old notebook `try_welsh_anova`
That also includes a `confint_median` function checked with R DescTools MedianCI. Not sure whether I already put this somewhere.

According to that notebook, the following is similar Bartlett test for equality of variances, but pvalue doesn't match exactly, maybe F versus chisquare problem
`anova_scale(arrs, method=method, center="mean", transform="square")[0]`
method='equal` (or originally "levene" comes closest

aside: I don't remember if I have this already somewhere. 
Under the null of equal variance and given normal distribution, the 4th moment are also constant, so "equal" would be justified by this. 
But we don't use that variance of variance is sigma^4 under normality.

josef-pkt commented:
transformation for oneway scale test:

abs is in Levene BF
square is standard normal model
others: 
exp I don't remember from where I got the idea of adding `exp`. I will remove it again, but I now allow for a user provided transform function.
log x*2 : I'm using this as test case for user provided function

brief search

Md. Khairul Islam TRANSFORMED TESTS FOR HOMOGENEITY OF VARIANCES AND MEANS
PhD thesis, compares several transforms including log-square and sqrt-abs for Levene type, and some transformation for Bartlett type tests
https://etd.ohiolink.edu/!etd.send_file?accession=bgsu1150727264&disposition=inline

aside: this is now generic anova with transformed data. testing variance is just a special case.
We can do standard anova using identity transform, or anova with arbitrary other transforms.
E.g. we could use variance stabilizing transform for non-normal data.
This needs checking for reuse, but should be easy. (transform of proportions or poisson counts)

Also, center can no be a number. Trivial case is setting it to `0.`, i.e. don't center. There might be other usecases if we have already given estimates or a null hypothesis of the center or mean.

josef-pkt commented:
`TrimmedMean` class has almost complete unit test coverage, but needs docstrings and maybe cached properties.

There are still some workflow problems, when I want to do more with oneway analysis.
e.g. I have an OLS for two sample case, How do I get effectsize and non-centrality for the equivalent anova_oneway analysis?
It's not "obvious" how to get everything I want or need.

side objective: I'm trying to get noncentrality and requirements for power analysis out of (OLS) Results instance from models and compare with standalone `stats` analysis.

josef-pkt commented:
and I'm getting confused again: When do we divide by `nobs` and when by `df_resid` to compute `f2` effect size and noncentrality index?

for power we use scaled nc / nobs, but f2 from f_stat is dividing by df_resid.
FtestAnovaPower computes df_resid from nobs and uses that for df, but uses nobs to get noncentrality
`pow_ = stats.ncf.sf(crit, df_denom, df_num, effect_size**2 * nobs)`

also a bit annoying: `confint_effectsize_oneway`  does not include the estimated effect size, only confint

josef-pkt commented:
inconsistency between two functions
(I think the problem is in the literature, `_fstat2effectsize` was written based on the appendix of a different article than what I used for power and similar)

`effectsize_oneway` was written to target power computation, and corresponds now also to my draft for effectsize/noncentrality computation for OLS

```
smo._fstat2effectsize(ta.statistic, *ta.df)
<class 'statsmodels.tools.testing.Holder'>
f2 = 0.3427001684846128
eta2 = 0.2552320886884138
omega2 = 0.11579441565136984
eps2 = 0.20203438073758623
eps2_ = 0.2020343807375862
omega2_ = 0.19851421368305308

es_f2 = smo.effectsize_oneway(means, vars_, nobs, use_var="equal")  # at estimated params
es_f2
0.31290015383377695

es_f2 * nobs_t / ta.df_denom
0.3427001684846128
```

josef-pkt commented:
`_fstat2effectsize` is still missing unit tests, it doesn't fully match the rest, and might need unit tests based on `anova_lm`
otherwise only individual lines some options are uncovered

josef-pkt commented:
It looks messy to validate my confint for effect size with R
package `effect size` uses a different way to compute confint and I'm not sure that's correct. 
I can replicate it using an fstat implied by the nc confint.

package MBESS produces the same eta2 confint as I have.
`ci.pvaf(45.8, 3, 35, N=39, conf.level = 0.9)`

My original unit test verify with an example from Steiger, but that covers only parts of confint function

josef-pkt commented:
I had forgotten about Stata.
stata for eta2 matches my confint at around 5 decimals and also MBESS

`esizei 2 40 7.47403193349075, level(90)`

josef-pkt commented:
robust_descriptive.py is only added in the first commit, no changes to it in other commits
diff of commits without first: 
https://github.com/statsmodels/statsmodels/pull/6526/files/077c07fe0c98b1962b5896d6afe7ab7676cddce9..7f127f0ecad5f27308425c95726d7c771a35cc01

so I can just drop that commit to remove it

josef-pkt commented:
vectorized for adding effectsize and confint to, e.g., `anova_lm` result

`r = _fstat2effectsize(res_f["F"], df1, df2)` works fine with pandas, no asarray, so everything remains in pandas



confint_effectsize_oneway breaks with shape mismatch when calling special function for ncF inversion.

requires broadcasting with column vector, and numpy broad casting and extra dimension for confint bounds will not easily work with pandas.
need asarray, maybe split `ci = ncfdtrinc(df1, df2, [1 - alpha1s, alpha1s], f_stat)` into low and upp.

```
confint_effectsize_oneway(res_f["F"].values[:, None], df1[:,None], df2 * np.ones(len(df1))[:, None], alpha=0.05)
m:\josef_new\eclipse_ws\statsmodels\statsmodels_py37\statsmodels\statsmodels\stats\oneway.py:194: RuntimeWarning: divide by zero encountered in true_divide
  eta2 = 1 / (1 + 1 / f2)
<class 'statsmodels.tools.testing.Holder'>
f2 = [[0.         0.29686364]
 [0.18547894 1.05552203]
 [0.         0.13313091]]
eta2 = [[0.         0.22890891]
 [0.15645908 0.51350558]
 [0.         0.11748943]]
ci_omega2 = [[-0.01818182  0.2116078 ]
 [ 0.12329088  0.4910632 ]
 [-0.03636364  0.08344021]]
ci_nc = [[ 0.         16.62436397]
 [10.57229986 60.16475591]
 [ 0.          7.58846163]]
ci_f = [[0.         0.54485195]
 [0.43067267 1.02738602]
 [0.         0.36487108]]
ci_eta = [[0.         0.47844426]
 [0.39554909 0.71659304]
 [0.         0.34276731]]
ci_f_corrected = [[0.         0.77053701]
 [0.52746414 1.25828576]
 [0.         0.44687399]]
```

not clear to me:
anova_lm results table has only one `df` in each row. In type 3, the second df should be df_resid.
However, for type 1 and some type 2, the second df shouldn't be the full model df_resid, AFAIU.

josef-pkt commented:
a bit of sightseeing on github

two applications with anova
https://github.com/CancerRxGene/gdsctools/blob/master/gdsctools/anova.py 
a lot of application specific code around anova

https://github.com/sunshineDrizzle/CommonTools/blob/master/commontool/stats.py
a nice wrapper around anova_lm using formulas for oneway and two-way anova, add eta_squared and omega_squared as effect sizes to the anova table
(no license but straightforward usage of statsmodels.)

josef-pkt commented:
temporary storage, for udocstring

```
>>> from statsmodels.stats.oneway import effectsize_oneway
>>> from statsmodels.stats.power import FTestAnovaPower
>>>
>>> nobs = np.array([10, 12, 13, 15])
>>> delta = 0.5
>>> means_alt = np.array([-1, 0, 0, 1]) * delta
>>> vars_ = np.arange(1, len(means_alt) + 1)
>>>
>>> f2_alt = effectsize_oneway(means_alt, vars_, nobs, use_var="equal")
>>> f2_alt
0.04581300813008131
>>>
>>> kwds = {'effect_size': np.sqrt(f2_alt), 'nobs': 100, 'alpha': 0.05, 'k_group
s': 4}
>>> power = FTestAnovaPower().power(**kwds)
>>> power
0.39165892158983273
>>>
>>> f2_alt = effectsize_oneway(means_alt, vars_, nobs, use_var="unequal")
>>> f2_alt
0.060640138408304504
>>>
>>> kwds['effect_size'] = np.sqrt(f2_alt)
>>> power = FTestAnovaPower().power(**kwds)
>>> power
0.5047366512800622
>>>
>>> f2_alt = effectsize_oneway(means_alt, vars_, nobs, use_var="bf")
>>> f2_alt
0.04391324307956788
>>>
>>> kwds['effect_size'] = np.sqrt(f2_alt)
>>> power = FTestAnovaPower().power(**kwds)
>>> power
0.3765792117047725
```

josef-pkt commented:
the first commit with robust_descriptive is now in PR #6790

The oneway part continues with cleanup in PR #6789 

This PR is now only for the record with background information about methods and design decisions.
